### PR TITLE
Refactor FfiCall a bit

### PR DESCRIPTION
**Pull request by johnynek**
please take a look @snoble and let me know if you see any problems with this approach.

### CODE REVIEW

File: core/src/main/scala/org/bykn/bosatsu/LetFreeEvaluation.scala [hunk 1]
```diff
@@ -32,4 +32,12 @@ object LetFreeEvaluation {
 
   type Cache = Option[CMap[String, (Future[Value], Type)]]
   type ToLFV = Option[LetFreeValue => Future[Value]]
+
+
+  def exprFn(wrapper: (LetFreeEvaluation.LetFreeValue, rankn.Type, LetFreeEvaluation.Cache, LetFreeEvaluation.ToLFV) => Any): FfiCall = {
```
Comment (johnynek) [inline core/src/main/scala/org/bykn/bosatsu/LetFreeEvaluation.scala:37]: we move the function from Ffi here so we keep value agnostic of as much as possible.

File: core/src/main/scala/org/bykn/bosatsu/Externals.scala [hunk 2]
```diff
@@ -1,69 +1,5 @@
...
-          FnValue { e3 =>
-            fn(e1, e2, e3)
-          }
-        }
-      }
-
-    def call(t: rankn.Type): Value = evalFn
-  }
-
-  final case class ExprFn(wrapper: (LetFreeEvaluation.LetFreeValue, rankn.Type, LetFreeEvaluation.Cache, LetFreeEvaluation.ToLFV) => Any) extends FfiCall {
```
Comment (johnynek) [inline core/src/main/scala/org/bykn/bosatsu/Externals.scala:42]: this is moved to a function in LetFree

File: core/src/main/scala/org/bykn/bosatsu/Value.scala [hunk 3]
```diff
@@ -120,18 +120,20 @@ object Value {
...
 
-  trait FnValueArg {
-    def toFn: Value => Value
+  class FnValue(fnValueArg: FnValue.Arg) extends Value {
+    val arg: FnValue.Arg = fnValueArg
   }
 
-  case class SimpleFnValue(toFn: Value => Value) extends FnValueArg
+  object FnValue {
+    trait Arg {
```
Comment (johnynek) [inline core/src/main/scala/org/bykn/bosatsu/Value.scala:128]: minor nit: I prefer to use objects as namespaces than use longer names.

File: core/src/main/scala/org/bykn/bosatsu/FfiCall.scala [hunk 4]
```diff
@@ -0,0 +1,62 @@
+package org.bykn.bosatsu
+
+sealed abstract class FfiCall {
```
Comment (johnynek) [inline core/src/main/scala/org/bykn/bosatsu/FfiCall.scala:3]: move this to a separate file since I keep not finding it when I look for it.

**Comment by johnynek** [comment]
see #496 for related change.

File: core/src/main/scala/org/bykn/bosatsu/LetFreeEvaluation.scala (hunk 1 above)
Comment (snoble) [reply core/src/main/scala/org/bykn/bosatsu/LetFreeEvaluation.scala:37]: This is much nicer here

**Comment by snoble** [comment]
This feels much nicer and can probably accomodate future experimental evaluators better. Thank you!

**Comment by johnynek** [comment]
going to merge even though the patch isn't well covered (because it is staged to be covered by #513

**Pull request merged**